File: app/src/main/java/com/davenotdavid/dndheadlines/ArticleAdapter.java
package com.davenotdavid.dndheadlines;

import android.content.Context;
import android.support.annotation.NonNull;
import android.text.format.DateUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.List;

import static android.text.format.DateUtils.MINUTE_IN_MILLIS;

/**
 * Custom adapter, subclass of {@link ArrayAdapter}, that inflates views onto a ListView.
 */
public class ArticleAdapter extends ArrayAdapter<Article> {

    // Log tag constant.
    private static final String LOG_TAG = ArticleAdapter.class.getSimpleName();

    /**
     * Creates a {@link ArticleAdapter} object.
     *
     * @param context is the Activity context passed in.
     * @param articleList is the list of Articles passed in.
     */
    public ArticleAdapter(Context context, List<Article> articleList) {
        super(context, 0, articleList);
    }

    /**
     * Provides a view for an AdapterView (ListView, GridView, and etc.).
     *
     * @param position is the position in the list of data that should be displayed in the list
     *                 item view.
     * @param convertView is the recycled view to populate.
     * @param parent is the parent/root view group used for inflation.
     */
    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {

        // Initializes the following to hold child views later on.
        ViewHolder holder;

        // Checks if the existing view is being reused, inflates the view otherwise.
        if (convertView == null) {
            convertView =
                    LayoutInflater.from(getContext()).inflate(R.layout.list_item, parent, false);

            // Initializes the following child views for the sake of not initializing them
            // repeatedly.
            holder = new ViewHolder();
            holder.articleTitle = (TextView) convertView.findViewById(R.id.article_title);
            holder.publishTimeAgo = (TextView) convertView.findViewById(R.id.publish_time_ago);

            // Associates the holder with the view for later lookup.
            convertView.setTag(holder);
        } else {

            // Otherwise, the view already exists so retrieve it.
            holder = (ViewHolder) convertView.getTag();
        }

        // Initializes the following Article being viewed at the moment.
        Article currentArticle = getItem(position);

        // Sets the article's title as text with ranks (top 10). Otherwise (which is unlikely, but
        // better safe than sorry!), "<VOID>" is set.
        String rankText = String.valueOf(position + 1) + ". ";
        StringBuilder titleSb = new StringBuilder();
        titleSb.append(rankText);
        if (!currentArticle.getTitle().isEmpty()) {
            titleSb.append(currentArticle.getTitle());
        } else {
            titleSb.append("<VOID>");
        }
        holder.articleTitle.setText(titleSb.toString());

        // Initializes the following reference variable to render a view of how long ago the
        // article was published as of now (user's locale time). Otherwise, hides the view.
        String articlePubDateTime = currentArticle.getPublishedAt();
        if (!articlePubDateTime.equals("null")) { // JSON apparently returns a null string

            // DateTimeFormatter initialized that's used to retrieve the UTC (4 hours faster than
            // EST) time in milliseconds of when the article was published.
            DateTimeFormatter formatter = DateTimeFormat
                    .forPattern("yyyy-MM-dd HH:mm:ss")
                    .withZoneUTC();
            long millisecondsSinceUnixEpoch =
                    formatter.parseMillis(formatUTCDateTime(articlePubDateTime));

            // Uses one of DateUtils' static methods to compare how long ago the article was
            // published (e.g. 37min ago, 5hours ago, and etc.).
            CharSequence relativeTime = DateUtils.getRelativeTimeSpanString(
                    millisecondsSinceUnixEpoch,
                    System.currentTimeMillis(),
                    MINUTE_IN_MILLIS); // Minimum time to be displayed (secs would constitute as "0min ago")

            holder.publishTimeAgo.setText(relativeTime.toString());
        } else {
            holder.publishTimeAgo.setVisibility(View.INVISIBLE);
        }

        return convertView;
    }

    // ViewHolder class used to hold the set of child views.
    private static class ViewHolder {
        TextView articleTitle;
        TextView publishTimeAgo;
    }

    /**
     * Modifies the UTC date-time by removing the 'T' and 'Z' chars, and retrieving only the first
     * two chars of seconds (in case of partial seconds such as decimals) in order to satisfy
     * DateTimeFormat's format.
     *
     * @param utcDateTime is the UTC date-time according to ISO 8601 standards.
     */
    private String formatUTCDateTime(String utcDateTime) {
        StringBuilder dateTimeSb = new StringBuilder(utcDateTime);
        dateTimeSb.deleteCharAt(utcDateTime.length() - 1);
        String dateTime = dateTimeSb.toString();
        String formattedDateTime = dateTime.replace('T', ' ');
        String[] dateTimeParts = formattedDateTime.split(" ");
        String[] timeParts = dateTimeParts[1].split(":");
        String time = timeParts[0] + ":" +
                timeParts[1] + ":" +
                String.valueOf(timeParts[2].charAt(0)) +
                String.valueOf(timeParts[2].charAt(1));
        return dateTimeParts[0] + " " + time;
    }
}

File: app/src/main/java/com/davenotdavid/dndheadlines/ArticleAdapter.kt
package com.davenotdavid.dndheadlines

import android.support.v7.widget.RecyclerView
import android.text.format.DateUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import org.joda.time.format.DateTimeFormat

import android.text.format.DateUtils.MINUTE_IN_MILLIS

/**
 * Adapter subclass of [RecyclerView] that renders a list of Article data that's clickable via a
 * custom interface.
 */
class ArticleAdapter(private val mListItemClickListener: ListItemClickListener,
                     private val mArticleData: List<Article>) :
        RecyclerView.Adapter<ArticleAdapter.ArticleViewHolder>() {

    // Log tag constant.
    private val LOG_TAG = ArticleAdapter::class.java.simpleName

    // Interface for the list items that are clickable.
    interface ListItemClickListener {
        fun onListItemClick(article: Article)
    }

    override fun getItemCount(): Int {
        return mArticleData.size
    }

    /**
     * Inflates a layout for the list items.
     */
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ArticleViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ArticleViewHolder(inflater.inflate(R.layout.list_item, parent, false))
    }

    /**
     * Binds each view from the ViewHolder onto the list item layout, and ultimately the
     * RecyclerView.
     */
    override fun onBindViewHolder(holder: ArticleViewHolder, position: Int) {

        // Initializes the following Article being viewed at the moment.
        val currentArticle = mArticleData[position]

        // Sets the article's title as text with ranks (top 10 at most). Otherwise (which is
        // unlikely, but better safe than sorry!), "<VOID>" is set.
        val rankText = (position + 1).toString() + ". "
        val titleSb = StringBuilder()
        titleSb.append(rankText)
        if (currentArticle.title != null) {
            titleSb.append(currentArticle.title)
        } else {
            titleSb.append("<VOID>")
        }
        holder.mArticleTitleTV.text = titleSb.toString()

        // Initializes the following reference variable to render a view of how long ago the
        // article was published as of now (user's locale time). Otherwise, hides the view.
        val articlePubDateTime = currentArticle.publishedAt
        if (articlePubDateTime != null) {

            // Tries to format the publish time properly prior to setting it as text. Otherwise,
            // catches an exception and sets the text invisible.
            try {

                // DateTimeFormatter initialized that's used to retrieve the UTC (e.g. 4 hours
                // faster than EST) time in milliseconds of when the article was published.
                val formatter = DateTimeFormat
                        .forPattern("yyyy-MM-dd HH:mm:ss")
                        .withZoneUTC()
                val millisecondsSinceUnixEpoch =
                        formatter.parseMillis(formatUTCDateTime(articlePubDateTime))

                // Uses one of DateUtils' static methods to compare how long ago the article was
                // published (e.g. 37 minutes ago, 5 hours ago, and etc.).
                val relativeTime = DateUtils.getRelativeTimeSpanString(
                        millisecondsSinceUnixEpoch,
                        System.currentTimeMillis(),
                        MINUTE_IN_MILLIS) // Minimum time to be displayed (secs would constitute as "0min ago")

                // Initially converts relativeTime to a String to possibly set the following
                // TextView as "just now!". Or, the TextView will be hidden if the String is
                // something like "In 5 min" which sometimes appears due to a back-end bug.
                // Otherwise, sets the publish time as is.
                val relativeTimeString = relativeTime.toString()
                if (relativeTimeString == "0 minutes ago") {
                    holder.mPublishTimeTV.setText(R.string.just_now_text)
                } else if (relativeTimeString[0] == 'I' && relativeTimeString[1] == 'n') {
                    holder.mPublishTimeTV.visibility = View.INVISIBLE
                } else {
                    holder.mPublishTimeTV.text = relativeTimeString
                }
            } catch (e: IndexOutOfBoundsException) {
                Log.e(LOG_TAG, "Problem formatting time string", e)

                holder.mPublishTimeTV.visibility = View.INVISIBLE
            }
        } else {
            holder.mPublishTimeTV.visibility = View.INVISIBLE
        }
    }

    /**
     * RecyclerView's ViewHolder class.
     */
    inner class ArticleViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
            View.OnClickListener {
        val mArticleTitleTV: TextView = itemView.findViewById(R.id.article_title) as TextView
        val mPublishTimeTV: TextView = itemView.findViewById(R.id.publish_time_ago) as TextView

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View?) {
            mListItemClickListener.onListItemClick(mArticleData[adapterPosition])
        }
    }

    /**
     * Modifies the UTC date-time by removing the 'T' and 'Z' chars, and retrieving only the first
     * two chars of seconds (in case of partial seconds such as decimals) in order to satisfy
     * DateTimeFormat's format.
     *
     * @param utcDateTime is the UTC date-time according to ISO 8601 standards.
     *
     * @throws [IndexOutOfBoundsException] since some custom indexing will be out of bounds of the
     *                                     # of time-attribute parts.
     */
    @Throws(IndexOutOfBoundsException::class)
    private fun formatUTCDateTime(utcDateTime: String?): String {
        val dateTimeSb = StringBuilder(utcDateTime)
        dateTimeSb.deleteCharAt(utcDateTime!!.length - 1)
        val dateTime = dateTimeSb.toString()
        val formattedDateTime = dateTime.replace('T', ' ')
        val dateTimeParts = formattedDateTime.split(" ".toRegex()).dropLastWhile {
            it.isEmpty()
        }.toTypedArray()
        val timeParts = dateTimeParts[1].split(":".toRegex()).dropLastWhile {
            it.isEmpty()
        }.toTypedArray()
        val time = timeParts[0] + ":" +
                timeParts[1] + ":" +
                timeParts[2][0].toString() +
                timeParts[2][1].toString()

        return dateTimeParts[0] + " " + time
    }
}